cloud-aws/src/main/java/com/sequenceiq/cloudbreak/cloud/aws/AwsSubnetIgwExplorer.java

```diff
@@ -68,4 +86,140 @@ public boolean hasInternetGatewayOfSubnet(List<RouteTable> routeTables, String s
+                gatewayIds.addAll(subnetRouteTable.get().getRoutes().stream()
+                    .filter(route -> StringUtils.isNotEmpty(route.getGatewayId()) && route.getGatewayId().startsWith(FW_PREFIX))
+                    .map(Route::getGatewayId)
+                    .collect(Collectors.toList())));
+        return gatewayIds;
+    }
+
+    private List<FirewallMetadata> getAllFirewallMetadataForVpc(AmazonNetworkFirewallClient nfwClient, String vpcId) {
+        ListFirewallsRequest request = new ListFirewallsRequest()
+            .withVpcIds(vpcId);
+        return AwsPageCollector.getAllFirewallMetadata(nfwClient, request);
+    }
+
+    private boolean hasRouteToPublicFirewall(List<FirewallMetadata> firewallMetadata, List<RouteTable> routeTables,
+            String subnetId, String vpcId, AmazonNetworkFirewallClient nfwClient, List<String> gatewayIds) {
+        LOGGER.info("Searching firewalls for gateway id in subnet {} route table.", subnetId);
+        boolean result = false;
+        List<AwsFirewall> firewalls = getFirewalls(firewallMetadata, nfwClient);
+        List<SyncState> syncStates = getStateForRoutedFirewalls(firewalls, nfwClient, gatewayIds, subnetId);
+        if (!syncStates.isEmpty()) {
+            LOGGER.info("Found {} firewall status states that match gateway ids in the subnet {} route table",
+                syncStates.size(), subnetId);
+            Optional<SyncState> syncStateOptional = syncStates.stream()
+                .filter(syncState -> hasInternetGatewayOfSubnet(routeTables, syncState.getAttachment().getSubnetId(), vpcId))
+                .findFirst();
+            result = syncStateOptional.isPresent();
+            if (result) {
+                String firewallName = firewalls.stream()
+                    .filter(fw -> fw.getAllSyncStates().contains(syncStateOptional.get()) && fw.getName() != null)
+                    .map(AwsFirewall::getName)
+                    .findAny().orElse("[could not find name]");
+                LOGGER.info("Found routed firewall {} with attached internet gateway for subnet {}", firewallName, subnetId);
+            } else {
+                LOGGER.info("The firewalls used by subnet {} do not have an internet gateway attached.", subnetId);
+            }
+        } else {
+            LOGGER.info("Did not find any firewall metadata that matched the route table for subnet {}", subnetId);
+        }
+        return result;
+    }
+
+    private List<AwsFirewall> getFirewalls(List<FirewallMetadata> firewallMetadata, AmazonNetworkFirewallClient nfwClient) {
+        List<String> firewallArns = firewallMetadata.stream()
+            .map(FirewallMetadata::getFirewallArn)
+            .collect(Collectors.toList());
+        LOGGER.debug("Issuing DescribeFirewallRequest for firewalls with arns {}", firewallArns);
+        return firewallArns.stream()
+            .map(arn -> nfwClient.describeFirewall(new DescribeFirewallRequest().withFirewallArn(arn)))
```

Comment:
How many RPCs does this perform and do you think a bulk API might be useful (assuming one exists)?

Reply:
It will run this request for each firewall in the VPC. There's unfortunately a bit of a disconnect in the way the data's presented that means we need to make this call. The subnet's route table will have all gateway ids, including firewalls, and the ListFirewallsRequest will give us the firewall arn and names. But the firewall gateway id is only visible in the DescribeFirewallResult. So we can't tell if the gateway ids found in the subnet's route table are firewall or not until we do this call.

I couldn't find a bulk API to use. However, most customers will never hit this code. If order for this check to work, you need to add additional permissions to your default AWS credential. If you don't have those, you'll get an `ACCESS_DENIED` earlier in the code and it will just move on. So this shouldn't be a general performance hit, and will only be encountered by customers who specifically enable it by adding the additional permissions.

Reply:
Ok, thanks for checking.